tests/provisioning/provision_test.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import os
+import unittest
+import subprocess
+from shlex import split
+import json
+import urllib.request
+
+from google.cloud.container_v1.services import cluster_manager
+from google.cloud.trace_v1 import trace_service_client
+from google.cloud import error_reporting
+
+class TestGKECluster(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        """ Get the location of GKE cluster for later queries """
+        command = ('gcloud container clusters list --format="value(location)"')
+        result = subprocess.run(split(command), encoding='utf-8', capture_output=True)
+        cls.location = str(result.stdout.replace('\n', ''))
+        if cls.location == '':
+            raise Exception('GKE cluster is not provisioned!')
+        cls.name = 'projects/' + getProjectId() + '/locations/' + cls.location + '/clusters/stackdriver-sandbox'
+    
+    def testNodeMachineType(self):
+        """ Test if the machine type for the nodes is as specified """
+        client = cluster_manager.ClusterManagerClient()
+        cluster = client.get_cluster(name=self.__class__.name)
+        node_config = cluster.node_config
+        self.assertEqual(node_config.machine_type, 'n1-standard-2')
+
+    def testNumberOfNode(self):
+        """ Test if the number of nodes in the node pool is in the specified range """
+        client = cluster_manager.ClusterManagerClient()
+        cluster = client.get_cluster(name=self.__class__.name)
+        self.assertTrue(cluster.current_node_count == 4)
+    
+    def testStatusOfServices(self):
+        """ Test if all the service deployments are ready """
+        command = ("kubectl get deployment --all-namespaces -o json")
+        result = subprocess.run(split(command), encoding='utf-8', capture_output=True)
+        services = json.loads(result.stdout)
+        for service in services['items']:
+            self.assertTrue(int(service['status']['readyReplicas']) >= 1)
+    
+    def testReachOfFrontend(self):
+        """ Test if query frontend returns 200 """
+        command_ip = ("kubectl get svc frontend-external -o json")
```

Comment:
We install Istio on the cluster, so getting the external IP of the cluster is not by "kubectl get svc frontend-external". Look at Nic's PRs for adding istio - there is a different command that returns external IP

Reply:
Here is the command for that:
kubectl -n istio-system get service istio-ingressgateway -o jsonpath='{.status.loadBalancer.ingress[0].ip}'

Reply:
How should I test the reachablity? I can't open the url for istio, but I can open the url for the frontend.